Rendering performance optimizations

dhruvachakravarthi opened the pull request:
### Main changes in this PR:
- StaticCanvasMarcher renamed to Endpoint
- Instead of continuously removing and adding fabric objects for pathways on every render, persistent objects are associated with every marcher_id
- These objects are grouped within a new class, MarcherVisualGroup. Every marcher visual group contains a CanvasMarcher, two Pathway(s), two Midpoint(s), and two Endpoint(s)
- These visual groups are accessible within a MarcherVisualMap, which maps marcher_id(s) to their MarcherVisualGroup
- A store exists for retrieving the MarcherVisualMap. This store is combined with the Marcher store, and useMarchersWithVisuals should be used when visuals are needed.
- Rendering path visuals for marchers has been optimized, which fixes the audio/animation stutter (although not foolproof by any means).
- Rendering temporary path visuals while dragging marchers is greatly optimized, and now easily handles batch editing 150+ marchers with both pathways on.
- Zooming has been optimized by caching most objects on the canvas, not only the grid
- Functions which traditionally accept MarcherPages but only require coordinate fields (.x and .y) now use the CoordinateLike type, located in CoordinateActions.ts
- Tests and mocks have been updated to conform to the new rendering functions and structures
### Videos:

https://github.com/user-attachments/assets/9b757f6b-6360-4f7c-92c0-96a6e62c9ff0

https://github.com/user-attachments/assets/0a57ae01-6627-4230-9c33-dc55e1a28819

https://github.com/user-attachments/assets/d58be9c8-09bc-4257-9eb9-6e3339e93023

https://github.com/user-attachments/assets/c52d7593-2019-4cf5-b6c5-fe2585a3d99c

AlexDumo commented:
Looking at the videos, this is fantastic! Lots of very lazy-loading was done by me on the canvas, so these optimizations are amazing

> Functions which traditionally accept MarcherPages but only require coordinate fields (.x and .y) now use the CoordinateLike type, located in CoordinateActions.ts

I've been wanting to do this for a while as we start to add props and sticky notes

> Rendering path visuals for marchers has been optimized, which fixes the audio/animation stutter (although not foolproof by any means).

I don't doubt this optimization will help, but the real issue with this is that the animation are based off of callback timers that just _hope_ that they'll line up with the music. The real solution to this (which I'm working on in a separate branch) is to make the animation conform to the current timestamp via keyframe calculation. Once I'm done with the first pass of this, I'll definitely want your help optimizing it. 

> These objects are grouped within a new class, MarcherVisualGroup. Every marcher visual group contains a CanvasMarcher, two Pathway(s), two Midpoint(s), and two Endpoint(s)

Very smart

> Zooming has been optimized by caching most objects on the canvas, not only the grid

Fixes #122?

AlexDumo commented:
Awesome! Any idea what's causing those 4 tests to fail? I'd like you to give it a look since you know more about this new structure

dhruvachakravarthi commented:
Yes, taking a look right now

dhruvachakravarthi commented:
@AlexDumo not sure what's happening here. GitHub actions reports:
On main:     Test Files  4 failed | 50 passed | 7 skipped (61)
On branch:  Test Files 4 failed | 50 passed | 7 skipped (61)

When running locally, 3/4 tests are skipped on main but run and fail on the branch. The one test that does fail is labelled:
StaticMarcherShape › marchers › should generate correct number of marcher objects

I can't find any test with the same description, so I'm not sure where this is coming from. Regardless, on GitHub actions the same number of tests fails and passes, so I'm not too worried.

AlexDumo commented:
@dhruvachakravarthi I lazily fixed all the tests (we were skipping them already anyway)

I'm ready to merge, but I'd like you to take a look at the bugs the cursor bot posted. Any thoughts?

dhruvachakravarthi commented:
@AlexDumo yes, will take a look later today.

dhruvachakravarthi commented:
@AlexDumo I just fixed bugs 2,3,4. Bug 1 shouldn't cause any issues. A side effect of the new way we update the store is that section appearances are updated without the need to refresh, so I've removed the prompt to refresh and automatic refresh when saving changes.
